Improve user friendliness of error for listing all members in a data set

namanpatel1112 opened the pull request:
Resolves https://github.com/zowe/zowe-cli/issues/935

zFernand0 commented on packages/zosfiles/src/methods/list/List.ts:
```diff
@@ -76,7 +77,18 @@ export class List {
+                // Isolate the main error message to display
+                const splitMessage: string[] = error.message.split("details:");
+
+                // Throw short error message
+                process.stdout.write("\n--> Use option `--debug-response` or `--dr` with the command to see REST API level debugging info.\n\n");
+                throw Error(splitMessage[splitMessage.length - 1].replace(/\s+$/, ""));
```
I believe that we want to avoid writing directly to process.stdout (or stderr) directly.
We should always aim to use the `params.response.console` functions to do so.
However, there is no such concept as a CommandResponse object here since this is an SDK method.

Given that we are throwing an error, maybe we could add the short error message in the `ImperativeError.additionalDetails` property (or some other `ImperativeError.shortDescription` optional property) and let the receiving handler react to such error and print the short message if available.

kudos to @t1m0thyj for helping come up with idea 😋

dkelosky commented:
Can we see some samples of the more friendly messages pasted here?

namanpatel1112 commented:
**Without using the --dr or --debug-response option:**
![image](https://user-images.githubusercontent.com/17503994/127057914-63526af0-5cff-4872-a2de-c46c82290b5e.png)
In the above, if we can remove the 'Unexpected Command Error' and 'Stack' trace messages and just show the 'Message' part, it would be even better. I am not sure how to prevent it from showing the unexpected command error and the stack trace. Let me know if there is a way.


**Using the --dr or --debug-response option, you can see all the REST API info:**
![image](https://user-images.githubusercontent.com/17503994/127058041-6386217e-f177-4b15-89e1-5479c74a989a.png)

zFernand0 commented:
Should we reserve `--dr` for dry-run? #870

namanpatel1112 commented:
Oh Okay, I wasn't aware --dr is being used for dry-run. So, do you think we can use --message-Response or --mr as Joe Winchester suggested in the issue related to this PR or any other name?

zFernand0 commented:
> Oh Okay, I wasn't aware --dr is being used for dry-run. So, do you think we can use --message-Response or --mr as Joe Winchester suggested in the issue related to this PR or any other name?

`--dry-run` has not been implemented as a global option yet. But we may want to use something else for this option.
Unless we think we should use full option names for global ones (i.e. no aliases).
Thank you for considering using other option names like `--message-response | --mr`. 😋

Joe-Winchester commented:
@namanpatel1112 - what's the best way I can test this PR locally (i.e. get it installed on my mac ? 

Looking at the screen shots if there is a basic 
```
zowe files list all-members "HLQ.NOTHERE"
```
Does this then write the message 
```
--> Use option `--debug-response` ... 
```
if the option hasn't been added ?  

I think as a user the response to 
```
zowe files list ds "HLQ.NOTHERE.*
```
would be
```
ISRZ002 Data set not cataloged - 'HLQ.NOTHERE' was not found in catalog
```
and if a user did 
```
zowe files list ds "HLQ.NOTHERE.* -h
```
then list that the option about getting more detailed information, i.e. make that only echo'd if the user asks for it

For other commands that fall into the same category when I use them to access files that aren't there and get a big gnarly error messages that your more verbose response would benefit from are:
```
zowe files list uss "/nothere"
zowe files upload file-to-uss "file.txt" "/nothere/file.txt"
zowe files download uss "/nothere"
zowe files download ds "HLQ.PDS(NOTHERE)
zowe files upload dir-to-pds "localpath" "HLQ.PDS(NOTHERE)"
```
Generally I find it's the file APIs that are the culprits.  There are probably others than the list shown above.

namanpatel1112 commented:
@Joe-Winchester Yes, you can install it on your local machine. But, hold off on that for now. I think your point about only showing the message '--> Use option `--debug-response...` when the user adds a help option is good. @zFernand0 What do you think? 

If I do this, then we don't have to go into ImperativeError.additionalDetails or create a ImperativeError.shortDescription property, which keeps it much more simple.

Joe, you can test it on your mac after I make the recommended changes.

zFernand0 commented:
> showing the message "--> Use option --debug-response..." when the user adds a help option is good.> 
> If I do this, then we don't have to go into ImperativeError.additionalDetails or create a ImperativeError.shortDescription property, which keeps it much more simple.

It definitely sounds simpler.
Will this message (`Use option --mr`) show up on all errors or just the errors on the commands where the  `--mr` option is supported?

side-comment:
- Should we have a bot or some sort of automation that does the `npm update` for us periodically?

namanpatel1112 commented:
@zFernand0 I think I might have misunderstood you about using ImperativeError.additional details property. I thought I would have to make changes in the imperative framework itself, but I just had to actually throw an imperative error with the error message in the additionalDetails property. So, I implemented it as you recommended.

**Now, without the --message-response | --mr option:**
![image](https://user-images.githubusercontent.com/17503994/127233981-1e06afdb-4fca-4adf-938e-be2977453a83.png)

**With the --message-response | --mr option:**
![image](https://user-images.githubusercontent.com/17503994/127234088-65a15fba-dafc-4cf2-8f92-b32802c49474.png)


The message (Use option --mr) doesn't show up with the error output itself to keep the error clutter free as Joe recommended. Instead I added as an option in the --help doc. So, users can see it under OPTIONS when they check the help doc (zowe files list am "naman.*" -h). So, it would only show up for commands that have that option.
![image](https://user-images.githubusercontent.com/17503994/127233866-58821b99-eeae-48fc-a1b2-b9b7e0da8c90.png)

I am just working on fixing the issue of codecov check failing, and then, I can do the same thing for other zosfiles commands:
  zowe files list uss "/nothere"
  zowe files upload file-to-uss "file.txt" "/nothere/file.txt"
  zowe files download uss "/nothere"
  zowe files download ds "HLQ.PDS(NOTHERE)
  zowe files upload dir-to-pds "localpath" "HLQ.PDS(NOTHERE)"

I think we should have a way to do `npm update` automatically periodically.

zFernand0 commented:
I'm a little worried about changing the default behavior of error messages.
Not sure if someone is depending on the REST output today to determine and react to errors in an automated pipeline.
Do we know if the `--response-format-json` output of the error is affected by us overwriting the error details?

namanpatel1112 commented:
That's true, I agree with that. Someone depending on the REST output would have to add the --mr flag to ensure there are no breaks in their automated pipeline. It does affect the --response-format-json output. I need to do it differently.

zFernand0 commented:
Reasons for closing this PR:
- It is a little bit behind the `master` branch.
- The squad has come to an agreement on how to move forward with improving error friendliness
  - See: https://github.com/zowe/zowe-cli/issues/935#issuecomment-1276179160
- This approach requires us to add our interpretation in every API where there is an error thrown.
